### ISSUE TITLE
Add Container (Docker) Image

### ISSUE THREAD

**Issue by jacque006** [thehubbleproject/hubble-contracts#561]
### What's wrong

Currently, we do not have an easy way to distribute a Hubble client for use by others. They would need to clone this repo and run it.

In addition, we run most commands through `ts-node`, which is not going to be as efficient as transpiling with typescript to JS and running with node.

### How can we fix it

- Add a `Dockerfile` and `.dockerignore` that supports building a typescript/nodejs based image.
- Add a Github Actions workflow which builds & pushes the image to a registry.

### Out of scope
- Integration/E2E tests of image w/ contracts

### Questions
- What container registry do we want to use? Docker Hub, Github Packages, etc.
- Should we use NodeJS v10 as specified in CI, or can we go to current LTS of v14?
- Can we use a `node-alpine` or another minimal image base, or do we need additional OS packages/deps for crypto or other operations?
- Do we want to push up an image on every commit? Or only for merges to `master` and releases?
- How do we want people to pass configuration to the container? Env vars, a config file (such as in https://github.com/thehubbleproject/hubble-contracts/issues/557).

**Comment by ChihChengLiang** [comment]
The followings are the concerns I've been avoiding building the Docker image.

- How do we make sure the image will correctly build? It's common that people having a Dockerfile in the codebase that doesn't build. I assume this concern is addressed since the github action will fail?
- We need extra effort to maintain two build environment (CI and the Docker image)
- The Hubble client will depend on many external files. How should we mount those files to the container's volume? Dependencies are
  - a genesis.json. We currently need to pass a genesis file. If we have a deployed testnet we can hardcode those. But some user might want to run a customized network so we should offer the option to mount a genesis.json
  - db directory. When we add a real database, we need a db path for that. XDG Base Directories is one of the pathes I'm considering to be the default db path.
  - client configs like the tokenID to feeReceiverID mapping.
  - eth1 keys. The coordinator has accounts that stakes, bids, and pays gas. We need to manage those key files.

**Comment by ChihChengLiang** [comment]
> What container registry do we want to use? Docker Hub, Github Packages, etc.

I've been using Docker Hub before 2020. No experience since. Either sounds good to me.

> Should we use NodeJS v10 as specified in CI, or can we go to current LTS of v14?

I've meant to upgrade to v14. The v14 has native bigInt support, we can consider using bigInt everywhere in the codebase so that we don't need to worry about MAX_SAFE_INTEGER or fractions anymore. I believe the original blockers to the upgrade, probably truffle or other js libs, are no longer exist.

> Can we use a node-alpine or another minimal image base, or do we need additional OS packages/deps for crypto or other operations?

Yeah, let's do as lightweight as possible

> Do we want to push up an image on every commit? Or only for merges to master and releases?

I would say merges to master and releases. The container registry has come quotas and frequently wait the image to build per commit seems not fun.

> How do we want people to pass configuration to the container? Env vars, a config file (such as in #557).

Ahhhh, I expanded this in the above comment.

**Comment by jacque006** [comment]
>     * How do we make sure the image will correctly build? It's common that people having a Dockerfile in the codebase that doesn't build. I assume this concern is addressed since the github action will fail?

Yes, if a step in our `Dockerfile` fails then the whole step will fail. I've used https://github.com/docker/build-push-action and can highly recommend it.

In addition, we can add a simple verification test that runs a container with the newly built image and pings/HTTP GETs to ensure it will start up.

>     * We need extra effort to maintain two build environment (CI and the Docker image)

Yes, there will be some added overhead with adding a Docker pipeline but I don't think it will be too bad. The main thing that will break it will be structural changes in our node or project.

>     * The Hubble client will depend on many external files. How should we mount those files to the container's volume? Dependencies are
>       
>       * a genesis.json. We currently need to pass a genesis file. If we have a deployed testnet we can hardcode those. But some user might want to run a customized network so we should offer the option to mount a genesis.json
>       * db directory. When we add a real database, we need a db path for that. XDG Base Directories is one of the pathes I'm considering to be the default db path.
>       * client configs like the tokenID to feeReceiverID mapping.
>       * eth1 keys. The coordinator has accounts that stakes, bids, and pays gas. We need to manage those key files.

Users of the image would need to specify those files as [volume mounts](https://docs.docker.com/storage/volumes/), and we of course would want to document them. For simpler setups `docker-compose` should be fine and it has [volume mounting options](https://docs.docker.com/compose/compose-file/compose-file-v3/#volumes).

For more production hardened setups such as in a container orchestrator like Kubernetes (k8s), those volume mount points would be exactly the same, although we may recommend different levels of security or persistence:
- `genesis.json` and client configs can be mounted via [Config Map](https://kubernetes.io/docs/concepts/configuration/configmap/#using-configmaps-as-files-from-a-pod)
- db directory can mount a [Persistent Volume Claim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#claims-as-volumes) which generally attach to more durable storage.
- eth1 keys (assuming they contain secrets/private keys) can be mounted from [Secrets](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod)

Further into the future we could even provide k8s manifests or a [Helm](https://helm.sh/) [Chart](https://artifacthub.io/) that make it easier to setup. But we have much bigger fish to fry first :)

**Comment by jacque006** [comment]
> I would say merges to master and releases. The container registry has come quotas and frequently wait the image to build per commit seems not fun.
It generally is not :) However, we can run the Docker build in parallel to other testing steps and it should finish before them if we wanted to build and/or push per commit/PR

**Comment by ChihChengLiang** [comment]
The file mounting looks so cool and super solid. 

I'll defer to you for whether to push the image on every commit

Looks like all issues addressed. We are ready to go Docker?

**Comment by jacque006** [comment]
![](https://media2.giphy.com/media/qhhamrBnxSKNG/giphy.gif)

**Comment by jacque006** [comment]
While working on this, also consider parallelizing parts of the NodeJS Github Actions workflow. For example, running the docker build and push, lint checks and fast tests, and then slower test suites can all be run side by side.
